streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -1006,14 +1009,21 @@ private void addTasksToStateUpdater() {
     }
 
     private void addTaskToStateUpdater(final Task task) {
+        final long nowMs = System.currentTimeMillis();
         try {
-            task.initializeIfNeeded();
-            stateUpdater.add(task);
+            if (canTryLock(task.id(), nowMs)) {
+                task.initializeIfNeeded();
+                stateUpdater.add(task);
+                taskIdToBackoffRecord.remove(task.id());
+            } else {
+                log.trace("Task {} is still not allowed to retry acquiring the state directory lock", task.id());
+                handleUnsuccessfulLockAcquiring(task, nowMs);
```

Comment:
Is this correct?
Every time initialization is attempted before the back-off, the time of the last attempt is updated to the current time. If we assume an attempt every poll interval and the poll interval is less than the back-off time, the task will never be initialized.
Assume the last unsuccessful attempt occurred at time 200 and now the current call to `canTryLock()` is 100ms later at time 300. Furthermore, assume the current back-off is 250. That is, `canTryLock()` should return `false` because 300 - 200 >= 250 is not `true`. The last attempt is updated to 300 and the backoff is exponentially updated with the increased number of attempt (let's say 500). If you try again in 100ms at 400 `canTryLock()` will again return `false`, because 400 - 300 >= 500 is still not true and it will also not be true next time. You should only update the back-off record if you actually have attempted to initialize the task and it was unsuccessful and not when you skipped the attempt due to the back-off.